types/react-datepicker/index.d.ts

```diff
@@ -35,7 +35,7 @@ interface HighlightDates {
     [className: string]: Date[];
 }
 
-export interface ReactDatePickerProps<Modifiers> {
+export interface ReactDatePickerProps<Modifiers = any> {
```

Comment:
To review this PR I researched how popperjs and react-popper uses modifiers.

One problem here is that the `Modifier` type accepts two type parameters (not only one) so I'm not sure how the current typing of `popperModifiers` will support type-checking of custom modifiers that use the second type parameter.  It looks like checking the second parameter is ignored, which is bad.

But a higher-level question is why do `ReactDatePicker` and `ReactDatePickerProps` need to be generic?  If the problem that #53588 was trying to solve was lack of custom-modifier support, would it be sufficient to change the type of `popperModifiers` to allow custom modifiers?  Like this:

```ts
  popperModifiers?: ReadonlyArray<Modifier<string, any>> | undefined;
```

If that solves the problem that #53588 was trying to solve, then I'd suggest removing the type parameters from `ReactDatePicker` and `ReactDatePickerProps` and essentially reverting to the pre-#53588 state other than changing the type of `popperModifiers` and adding a test for a custom and strict modifier. Doing this would have a few advantages over the code proposed in this PR:
* existing pre-#53588 code that used strict modifiers would continue to work without needing to add a type parameter when upgrading to `@types/react-datepicker@latest`. Avoiding breaking changes is preferred.
* Error messages and IDE inline hover-help (aka "Intellisense") would be simpler for these types without generics. This makes working with these types easier and less confusing for the majority of developers who don't use custom modifiers.
* It probably would make TS compilation slightly faster for developers who don't use custom modifiers.

Reply:
Thanks for the review

The react-popper `PopperProps` also does not have the second parameter for `Modifier`, so it should work the same way.
Your suggested change solves the problem of lack of type support for custom popper modifiers.
The main reason why I added a type parameter is better IntelliSense support when passing inline modifiers. The question is, are they worth it? Also, I don’t see a better approach to achieve this behavior.

Reply:
> The react-popper `PopperProps` also does not have the second parameter for `Modifier`, so it should work the same way.

Good point. It looks like react-popper's types just default the second type parameter to `object`, so any type checking is lost for passing options to modifiers. That seems not ideal. If we can solve that problem here, I could file an issue over at react-popper to fix types there.  IMHO, type safety is a higher priority than easy IntelliSense, but I suspect we may be able to figure out typings that achieve both goals. 

> better IntelliSense support when passing inline modifiers

Now I understand the appeal of the changes you made in #53588. With the types in this new PR, does IntelliSense still work in the same way?  And where does the desired IntelliSense actually show up?  In JSX for the modifiers prop of the react-datepicker component?  When declaring an object-literal props object?  Someplace else?

A code sample and/or screenshot showing how you're using these types and where the IntelliSense shows up would be helpful to clarify the use case.  Could you post?  In particular I'm curious about whether the IntelliSense is primarily for the names of modifiers, or also for options too. (If the latter then I must be misunderstanding something.)

Reply:
> It looks like react-popper's types just default the second type parameter to object, so any type checking is lost for passing options to modifiers. That seems not ideal.

Just to be sure we could suppress it with react-popper approach.

> A code sample and/or screenshot showing how you're using these types and where the IntelliSense shows up would be helpful to clarify the use case.

[CodeSandbox](https://codesandbox.io/s/datepicker-hints-jrshy?file=/src/App.tsx)

With create-react-app and typescript template the result is:
<details>
<summary>screenshots</summary>

![telegram-cloud-photo-size-2-5301097491879540027-y](https://user-images.githubusercontent.com/51246150/132053564-ddfd44ee-e172-4ca5-8cb8-c3402f8a1cfd.jpg)
![telegram-cloud-photo-size-2-5301097491879540028-y](https://user-images.githubusercontent.com/51246150/132053588-3d7ec33f-0537-471e-ba30-7b6b65ecdb73.jpg)
![image](https://user-images.githubusercontent.com/51246150/132053778-40135878-16cd-4539-a538-9c4cb4dc5640.png)
![image](https://user-images.githubusercontent.com/51246150/132054958-df867386-098d-4d1f-a325-35a7761ef694.png)
</details>

Reply:
Thanks, that repro code was very helpful. On CodeSandbox I didn't see the Intellisense nor did I see it in CRA, but I upgraded the CRA repo to TS 4.4 and finally I saw it working, and I can see why it's useful.  Interestingly, there's no TS errors about the missing type parameter when using the constructor in code or in JSX, but there is a type error shown when using the `ReactDatePickerProps` type. [Playground link](https://www.typescriptlang.org/play?&dtPR=#code/JYWwDg9gTgLgBAEQIYwKYAVgGMDWqoA0cA3nAEqpJYzJqa77pQRgDOcAvnAGbMhwByKJWoBaACYpUYbHigCA3AFgAUKEixyI+Lwj8BAOgD0AOwjjUAfRDmArgBtUrIwAEYATzBOjwqjEWqqlgQJqzwSHAAvHAmqADuiFL0cgAUpKpwmXAhAMIAFkgmAOaoAFxwKQCUUQB8JBxEcBlZkGBeUACy5sDcwPis5QDaxM1ZWSZIIGWCENzcrKj+o5wAuqoclcoqwaHwYMxs5RR+tBiyjAfs0ekqY7kFxdNVtfUEy63tXeI9fVADcMNlmMJlNygJZvNFgJlhw1ioOIFtiEwjEICwohVqpE6sQOAogA). Not sure why. Thankfully, this limits the scope of the breaking change from the previous PR to only developers who are using the props type explicitly.  That's great news.

I also now understand why there's type safety for strict modifier options because of the smart discriminated union type of `Modifier` in react-popper. There's still no type safety for custom modifier options, but IMHO that's OK for now. Someone else can improve these types later to make custom modifier typing better. :-)

A challenge with the current PR is that if you supply custom modifiers as the type parameter, then you'll lose IntelliSense and typing for strict modifiers unless you manually supply the names of all the strict modifiers too. What do you think about the approach below which would allow both custom modifier names and strict names to get typing and IntelliSense? 

```ts
type StrictModifierNames = NonNullable<Popper.StrictModifiers['name']>;

export interface ReactDatePickerProps<CustomModifierNames = never> {
. . .
    popperModifiers?: ReadonlyArray<Modifier<CustomModifierNames | StrictModifierNames>> | undefined;
. . .
} 

declare class ReactDatePicker<CustomModifierNames = never> extends
    React.Component<ReactDatePickerProps<CustomModifierNames>> {
. . .
```

Reply:
> On CodeSandbox I didn't see the Intellisense nor did I see it in CRA, but I upgraded the CRA repo to TS 4.4 and finally I saw it working, and I can see why it's useful.

Unfortunately, I also get this behavior on TS playground. However, I switched typescript version in my CRA, and type hints remains the same for both names and options:

<details>
<summary>screenshot</summary>

![telegram-cloud-photo-size-2-5303349291693224333-y](https://user-images.githubusercontent.com/51246150/132082408-724ff922-b182-4ed3-b3e5-ef5977615031.jpg)
</details>

> Interestingly, there's no TS errors about the missing type parameter when using the constructor in code or in JSX, but there is a type error shown when using the ReactDatePickerProps type.

As far as I can see, TypeScript doesn't show an error with a missing generic class parameter. Generic class infers type parameter from usage or, if it's impossible, there will be widening to `unknown`.

> What do you think about the approach below which would allow both custom modifier names and strict names to get typing and IntelliSense?

It seems good to me. Just to make it more type-safe, there should be string constrain:

```typescript
<CustomModifierNames extends string = never>
```

So, I'll try to reproduce missing type hints in different environments and if everything is good, I'll prepare a new commit with your suggestion.

Reply:
> What do you think about the approach below which would allow both custom modifier names and strict names to get typing and IntelliSense?

Existing code that uses custom modifiers will require a type argument for modifiers' names.

> A challenge with the current PR is that if you supply custom modifiers as the type parameter, then you'll lose IntelliSense and typing for strict modifiers unless you manually supply the names of all the strict modifiers too.

I think it's not that bad because the type argument is optional.

Maybe typing modifiers as `ReadonlyArray<Modifier<any, any>>` will be a good compromise? Using any as the first parameter gives the ability to suggest `CustomModifierNames`. The only disadvantage is weak type support for options because typescript can't infer a specific options set.

<details>
<summary>bad options</summary>

![telegram-cloud-photo-size-2-5307820610346399478-y](https://user-images.githubusercontent.com/51246150/132135038-f30d6eca-43c7-4578-822f-663e9d282f72.jpg)

</details>

Reply:
> Existing code that uses custom modifiers will require a type argument for modifiers' names.

This isn't ideal, but I think it's probably OK because: 
* Custom modifiers are less commonly-used than strict ones, so if there are tradeoffs then the best approach is usually to make things easier for the common case and possible (with work req'd) for the uncommon case.
* The solution will be easy to discover from looking at a TS error message that contains the word "CustomModifierNames".
* If the alternative is losing IntelliSense and type-safety for strict options, then I don't think that tradeoff is worth it.

> > A challenge with the current PR is that if you supply custom modifiers as the type parameter, then you'll lose IntelliSense and typing for strict modifiers unless you manually supply the names of all the strict modifiers too.
> 
> I think it's not that bad because the type argument is optional.

I think it'd be better to make custom modifiers additive instead of replacing strict modifiers. Otherwise, if a developer adds a single custom modifier then their existing code will fail to compile unless they manually add strict modifiers to the type argument.

> The only disadvantage is weak type support for options because typescript can't infer a specific options set.

I think this is a pretty big disadvantage. There are a lot of modifiers and knowing which options go with which (both at compile time and when editing with IntelliSense) is very helpful.

Reply:
Pushed a commit with suggested changes. Also created a [repo](https://github.com/tsvetkovda/datepicker-hints) to reproduce locally

Reply:
Looks good. Thanks!